scripts/storage/qcow2/managesnapshot.py

```diff
@@ -0,0 +1,279 @@
+        logging.info('Backuping snapshot %s to %s', src, dst)
+        if not os.path.isdir(backupdir):
+            os.makedirs(backupdir, 0777)
+        shutil.copyfile(src, dst)
+        logging.info('Snapshot backup completed %s to %s', src, dst)
+    except Exception as e:
+        logging.error('An error occured during snapshot backup')
+        logging.error('%s', e)
+        logging.error('Destroying snapshot')
+        destroy_snapshot()
+        raise
+
+
+def getdev(dom0, snapshot_file_path):
+    xmldesc = dom0.XMLDesc(0)
+    root = etree.fromstring(xmldesc)
+    for i in root.findall('./devices/disk'):
+        device = {}
+        if 'disk' in i.attrib.values():
+            child = i.getchildren()
+            for childitems in child:
+                if childitems.tag == 'source':
+                    device['source'] = (childitems.get('file'))
+                elif childitems.tag == 'target':
+                    device['dev'] = (childitems.get('dev'))
+        if snapshot_file_path in device.values():
+            dev = device.get('dev')
+            return dev
+
+
+def wait_for_block_job(dom0, dev, abort_on_error=False,
+                       wait_for_job_clean=False):
+        """Wait for libvirt block job to complete.
+        Libvirt may return either cur==end or an empty dict when
+        the job is complete, depending on whether the job has been
+        cleaned up by libvirt yet, or not.
+        :returns: True if still in progress
+                  False if completed
+        """
+
+        status = dom0.blockJobInfo(dev, 0)
+        if status == -1 and abort_on_error:
+            logging.error('libvirt error while requesting blockjob info')
+            raise libvirt.VIR_ERR_OPERATION_FAILED('libvirt error while requesting blockjob info')
+        try:
+            cur = status.get('cur', 0)
+            end = status.get('end', 0)
+        except Exception:
+            return False
+
+        if wait_for_job_clean:
+            job_ended = not status
+        else:
+            job_ended = cur == end
+
+        return not job_ended
+
+
+if __name__ == "__main__":
+    global createarg, deletearg, backuparg, backupdir, backup_file_name, disk_path, domain, snapshot_file_name, snapshot_file_path
+    args = getargs()
+    createarg = args['createarg']
+    destroyarg = args['destroyarg']
+    backuparg = args['backuparg']
+    backupdir = args['backupdir']
+    backup_file_name = args['backup_file_name']
+    disk_path = args['disk_path']
+    domain = args['domain']
+
+    snapshot_file_name = disk_path.split('/')[-1] + "-snap.qcow2"
+    path = disk_path.split('/')
+    del path[-1]
+    path = '/'.join(path)
+    snapshot_file_path = path + "/" + snapshot_file_name
+
+    logfile = "/var/log/cloudstack/agent/snapshot.log"
+    logging.basicConfig(format='%(asctime)s %(pathname)s %(levelname)s:%(message)s', level=logging.DEBUG, filename=logfile)
+
+    facts = facter.Facter(facter_path="/usr/local/bin/facter")
```

Comment:
Hard coding /usr/local/bin/facter everywhere will lead us to trouble if the path changes in the future. Maybe this should be passed as argument or environment variables.